extensions/typescript-language-features/src/languageFeatures/diagnostics.ts

```diff
@@ -148,6 +151,80 @@ class DiagnosticSettings {
 	}
 }
 
+class DiagnosticsTelemetryManager extends Disposable {
+
+	private readonly _diagnosticCodesMap = new Map<number, number>();
+	private readonly _diagnosticSnapshotsMap = new ResourceMap<readonly vscode.Diagnostic[]>(uri => uri.toString(), { onCaseInsensitiveFileSystem: false });
+	private _timeout: NodeJS.Timeout | undefined;
+
+	constructor(
+		private readonly _telemetryReporter: TelemetryReporter,
+		private readonly _diagnosticsCollection: vscode.DiagnosticCollection,
+	) {
+		super();
+		this._register(vscode.workspace.onDidChangeTextDocument(e => {
+			if (e.document.languageId === 'typescript' || e.document.languageId === 'typescriptreact') {
+				this._updateAllDiagnosticCodesAfterTimeout();
+			}
+		}));
+		this._updateAllDiagnosticCodesAfterTimeout();
+		this._registerTelemetryEventEmitter();
+	}
+
+	private _updateAllDiagnosticCodesAfterTimeout() {
+		clearTimeout(this._timeout);
+		this._timeout = setTimeout(() => this._updateDiagnosticCodes(), 5000);
+	}
+
+	private _increaseDiagnosticCodeCount(code: string | number | undefined) {
+		if (code === undefined) {
+			return;
+		}
+		this._diagnosticCodesMap.set(Number(code), (this._diagnosticCodesMap.get(Number(code)) || 0) + 1);
+	}
+
+	private _updateDiagnosticCodes() {
+		this._diagnosticsCollection.forEach((uri, diagnostics) => {
+			const previousDiagnostics = this._diagnosticSnapshotsMap.get(uri);
+			this._diagnosticSnapshotsMap.set(uri, diagnostics);
+			const diagnosticsDiff = diagnostics.filter((diagnostic) => !previousDiagnostics?.some((previousDiagnostic) => equals(diagnostic, previousDiagnostic)));
+			diagnosticsDiff.forEach((diagnostic) => {
+				const code = diagnostic.code;
+				this._increaseDiagnosticCodeCount(typeof code === 'string' || typeof code === 'number' ? code : code?.value);
+			});
+		});
+	}
+
+	private _registerTelemetryEventEmitter() {
+		setInterval(() => {
```

Comment:
Make sure to also clean up this in `dispose`

Reply:
That's right, thanks for noticing